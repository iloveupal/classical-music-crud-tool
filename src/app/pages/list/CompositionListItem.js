import React, { PureComponent } from "react";
import classnames from "classnames";
import PropTypes from "prop-types";

import { Card } from "antd";
import {
  getComposer,
  getMovements,
  getTitle
} from "domains/composition/propGetters";

import MovementListItem from "./MovementListItem";

import "./styles/list-item.less";

function renderTitle({ title, composer }) {
  return (
    <div className={"composition-list-item__card-title"}>
      <span
        className={classnames({
          "composition-list-item__title": !!title,
          "composition-list-item__title-empty": !title
        })}
      >
        {title || "No title"}
      </span>
      <span className={"composition-list-item__composer"}>{composer}</span>
    </div>
  );
}

function renderExtra({ onDelete, onEdit, id }) {
  return (
    <div>
      <span
        onClick={() => onEdit(id)}
        className={"composition-list-item__edit-link"}
      >
        Edit
      </span>
      <span
        onClick={() => onDelete(id)}
        className={"composition-list-item__delete-link"}
      >
        Delete
      </span>
    </div>
  );
}

class CompositionListItem extends PureComponent {
  render() {
    const { onDelete, onEdit } = this.props;
    const composition = this.props.data;

    const title = getTitle(composition);
    const composer = getComposer(composition);
    const movements = getMovements(composition);

    return (
      <Card
        title={renderTitle({ title, composer })}
        extra={renderExtra({ onDelete, onEdit, id: composition._id })}
        className={"composition-list-item"}
      >
        {(movements.length &&
          movements.map(movement => {
            return <MovementListItem key={movement._id} data={movement} />;
          })) ||
          null}
        {(!movements.length && (
          <span className={"composition-list-item__no-movements"}>
            No movements found
          </span>
        )) ||
          null}
      </Card>
    );
  }
}

CompositionListItem.propTypes = {
  data: PropTypes.object.isRequired,
  onDelete: PropTypes.func.isRequired,
  onEdit: PropTypes.func.isRequired
};

export default CompositionListItem;
