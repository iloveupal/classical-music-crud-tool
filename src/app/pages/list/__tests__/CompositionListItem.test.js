import React from "react";

import CompositionListItem from "../CompositionListItem";

import { shallow } from "enzyme";

const testComposition = {
  _id: "123",
  title: "Symphony",
  composer: "Beethoven",
  movements: [
    {
      _id: "245",
      title: "First",
      key: "c_s_m",
      recordings: [
        {
          _id: "456"
        }
      ]
    }
  ]
};

describe("CompositionListItem", () => {
  it("should match the snapshot", () => {
    const wrapper = shallow(
      <CompositionListItem
        data={testComposition}
        onDelete={jest.fn()}
        onEdit={jest.fn()}
      />
    );

    expect(wrapper).toMatchSnapshot();
  });
});
