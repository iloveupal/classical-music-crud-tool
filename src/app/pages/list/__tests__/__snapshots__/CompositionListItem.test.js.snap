// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CompositionListItem should match the snapshot 1`] = `
<Card
  className="composition-list-item"
  extra={
    <div>
      <ColorLink
        className="composition-list-item__edit-link"
        onClick={[Function]}
        type="primary"
      >
        Edit
      </ColorLink>
      <ColorLink
        onClick={[Function]}
        type="danger"
      >
        Delete
      </ColorLink>
    </div>
  }
  title={
    <div
      className="composition-list-item__card-title"
    >
      <span
        className="composition-list-item__title"
      >
        Symphony
      </span>
      <span
        className="composition-list-item__composer"
      >
        Beethoven
      </span>
    </div>
  }
>
  <MovementListItem
    data={
      Object {
        "_id": "245",
        "key": "c_s_m",
        "recordings": Array [
          Object {
            "_id": "456",
          },
        ],
        "title": "First",
      }
    }
    key="245"
  />
</Card>
`;
